lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart

```diff
@@ -0,0 +1,690 @@
+    // Search for the beginning of a @font-face.
+    if (!resolvingFontFace) {
+      if (line == '@font-face {') {
+        resolvingFontFace = true;
+      } else {
+        continue;
+      }
+    } else {
+      // We are resolving a @font-face, read out the url and ranges.
+      if (line.startsWith('  src:')) {
+        int urlStart = line.indexOf('url(');
+        if (urlStart == -1) {
+          throw new Exception('Unable to resolve Noto font URL: $line');
+        }
+        int urlEnd = line.indexOf(')');
+        fontFaceUrl = line.substring(urlStart + 4, urlEnd);
+      } else if (line.startsWith('  unicode-range:')) {
+        fontFaceUnicodeRanges = <_UnicodeRange>[];
+        String rangeString = line.substring(17, line.length - 1);
+        List<String> rawRanges = rangeString.split(', ');
+        for (final String rawRange in rawRanges) {
+          List<String> startEnd = rawRange.split('-');
+          if (startEnd.length == 1) {
+            String singleRange = startEnd.single;
+            assert(singleRange.startsWith('U+'));
+            int rangeValue = int.parse(singleRange.substring(2), radix: 16);
+            fontFaceUnicodeRanges.add(_UnicodeRange(rangeValue, rangeValue));
+          } else {
+            assert(startEnd.length == 2);
+            String startRange = startEnd[0];
+            String endRange = startEnd[1];
+            assert(startRange.startsWith('U+'));
+            int startValue = int.parse(startRange.substring(2), radix: 16);
+            int endValue = int.parse(endRange, radix: 16);
+            fontFaceUnicodeRanges.add(_UnicodeRange(startValue, endValue));
+          }
+        }
+      } else if (line == '}') {
+        subsets.add(
+            _ResolvedNotoSubset(fontFaceUrl!, name, fontFaceUnicodeRanges!));
+        resolvingFontFace = false;
+      } else {
+        continue;
+      }
+    }
+  }
+
+  return _ResolvedNotoFont(name, subsets);
+}
+
+void _registerResolvedFont(_NotoFont font, _ResolvedNotoFont resolvedFont) {
+  _resolvedNotoFonts[font] = resolvedFont;
+
+  for (_ResolvedNotoSubset subset in resolvedFont.subsets) {
+    for (_UnicodeRange range in subset.ranges) {
+      _resolvedNotoTreeRoot =
+          _insertNotoFontRange(range, subset, _resolvedNotoTreeRoot);
+    }
+  }
+}
+
+/// Finds the minimum set of fonts which covers all of the [codeunits].
+///
+/// Since set cover is NP-complete, we approximate using a greedy algorithm
+/// which finds the font which covers the most codeunits. If multiple CJK
+/// fonts match the same number of codeunits, we choose one based on the user's
+/// locale.
+Set<_NotoFont> _findMinimumFontsForCodeunits(
+    List<int> codeunits, Set<_NotoFont> fonts) {
+  List<int> unmatchedCodeunits = List<int>.from(codeunits);
+  Set<_NotoFont> minimumFonts = <_NotoFont>{};
+  List<_NotoFont> bestFonts = <_NotoFont>[];
+  int maxCodeunitsCovered = 0;
+
+  while (unmatchedCodeunits.isNotEmpty) {
+    for (var font in fonts) {
```

Comment:
Potential infinite loop. Add code+comment on what happens when an invalid code unit arrives that causes unmatchedCodeunits. Suggestion protect with a loop counter, if exhausted, log warning with list of codeunits that were unexpectedly not resolved.

Reply:
Fixed by ensuring that we only ever call this with code units which we have already determined are covered.